File: 0993-cousins-in-binary-tree/0993-cousins-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int xDepth = -1, yDepth = -2;
    TreeNode* xParent = NULL, *yParent = NULL;
    void dfs(TreeNode* root, int x, int y, TreeNode* parent, int depth) {
        if(!root) return;
        if(root->val == x) {
            xParent = parent;
            xDepth = depth;
        }
        if(root->val == y) {
            yParent = parent;
            yDepth = depth;
        }
        dfs(root->left, x, y, root, depth + 1);
        dfs(root->right, x, y, root, depth + 1);
    }
    bool isCousins(TreeNode* root, int x, int y) {
        dfs(root, x, y, NULL, 0);
        return xDepth == yDepth && xParent != yParent;
    }
};

File: 0993-cousins-in-binary-tree/0993-cousins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode xParent = null, yParent = null;
    int xHeight = 0, yHeight = 0;
    public boolean isCousins(TreeNode root, int x, int y) {
        solve(root, x, y, null, 0);
        return xHeight == yHeight && xParent.val != yParent.val;
    }
    void solve(TreeNode root, int x, int y, TreeNode parent, int depth) {
        if(root == null) return;
        if(root.val == x) {
            xParent = parent;
            xHeight = depth;
        }
        if(root.val == y) {
            yParent = parent;
            yHeight = depth;
        }
        solve(root.left, x, y, root, depth+1);
        solve(root.right, x, y, root, depth+1);
    }
}